design/client_conformance_profiles.md

```diff
@@ -0,0 +1,550 @@
+
+* Admin Console
+    - UI
+    - CLI
+    - REST API directly
+    - Import
+
+### Client's Request Check
+
+Keycloak needs to do Client's Request Check at the following points.
+
+* Authorization Request
+* Token Request
+* Token Refresh Request
+* Token Introspection Request
+* UserInfo Request
+* (Back Channel) Logout Request
+
+### Security Profile Specific Operation on the Client's Request
+
+Depending on applied security profile, keycloak needs to do Security Profile Specific Operation on the Client's Request at the following points.
+
+* Authorization Response
+* Token Response
+* Token Refresh Response
+* Token Introspection Response
+* UserInfo Response
+* (Back Channel) Logout Response
+
+### Client Conformance Profiles SPI
+
+Depending on applied security profile, keycloak needs to process Client's Configuration Check, Client's Request Check, and Security Profile Specific Operation on the Client's Request.
+
+To realize it, Client Conformance Profile SPI is introduced. Its provider implementations are also provided per client conformance profile. 
+
+## Implementation
+
+### Plan
+
+Ph 1. Introducing Client Conformance Profile Mechanism (SPI Provider)
+
+Ph 2. Supporting FAPI Read Only Security Profile (Implementing Provider)
+
+Ph 3. Supporting FAPI Read and Write Security Profile (Implementing Provider)
+
+Ph (TBD). Supporting OpenBanking Security Profile
+
+Ph (TBD). Supporting CDR Security Profile
+
+### FAPI Read Only API Security Profile
+
+#### Endpoint : Dynamic Client Registration Endpoint
+
+Client's Configuration Check
+
+* Operations :
+    - Registering the client
+    - Configuring the existing client
+
+##### 1. Client Authentication Method
+OpenID Connect Client Metadata - `token_endpoint_auth_method` is one of the following values.
+ 
+ 1. `tls_client_auth_subject_dn` (currently, keycloak only supports it for Mutual TLS for OAuth Client Authentication[MTLS])
+ 2. `client_secret_jwt`
+ 3. `private_key_jwt`
+
+* Reference:
+    - Authz Server / 4, 4-1, 4-2
+    - Confidential Client / 1, 1-1, 1-2
+
+Error Response : 400 Bad Request (the detail is TBD)
+
+##### 2. Proof Key for Conde Exchange [PKCE]
+Set the client's `code_challenge_method` to `S256` to enforce it [PKCE].
+
+* Reference:
+    - Authz Server / 7
+    - Public Client / 1, 2
+
+Error Response : -
+
+##### 3. Redirect URI
+OpenID Connect Client Metadata - `redirect_uris` includes at least one URI.
+
+All URIs in `redirect_uris` do not include any wildcard, only specifying one URI.
+
+The scheme of all URIs in `redirest_uris` are https.
+
+* Reference:
+    - Authz Server / 8
+    - Authz Server / 20
+
+Error Response : 400 Bad Request (the detail is TBD)
+
+#### Endpoint : Authorization Endpoint
+
+Client's Request Check
+
+* Operations :
+    - handle authorization request
+
+##### 1. Proof Key for Conde Exchange [PKCE]
+The authorization request includes `code_challenge` and `code_challenge_method`.
+
+`code_challenge_method` is `S256`.

```

Comment:
We already have the switch on the client to enforce S256. This is the case for some other things described here.

In general, question is, what to do with the "small" switches, which we already have and how to handle if their value conflict with the switch for the conformance profile? Few options:
1) Remove all the "small" switches and just replace them with the new FAPI switches. However this may be the issue if some clients want more fine-grained settings. For example some clients may want to enforce PKCE as S256, but still want to authenticate with simple "client_secret" method. Which means that switch this client entirely to FAPI profile will be too restrictive for that client. Due to this, it seems that removing the "small" switches is not an option?

2) If some client will be in the admin console updated to use FAPI profile, then we will automatically update all the "small" switches. For example if client is updating to use FAPI read-only profile, then we will automatically update the "PKCE enforce" switch to S256. But it looks that  we can't do such automatic update for everything? For example if admin wants to update client to use FAPI read-write profile, but client uses "client_secret" method, then automatically changing client authentication method won't work IMO. I think we may need to throw the validation error like "Cannot change client to use FAPI read-write. Please update the client authentication method before" etc.

Those are the implementation details, but probably worth to discuss before we start implementing this?

Reply:
As for 1, IMO, we need to keep "PKCE enforce" switch. PKCE is not a part of FAPI Security Profile Specification. Some user wants to use "PKCE enforce" feature standalone.

As for 2, I think that it cost a lot for us to implement auto-updating feature about client settings. 

Therefore, as you said, it might be better to implement validation and throwing error when the admin updates FAPI Security Profile setting switch on Admin Console.

At this time, if the admin wants to change FAPI Security Profile setting, 

* He needs to do in advance appropriate setting for each client's setting items in comply with FAPI Security Profile setting to be enabled.
* After that, he changes the FAPI Security Profile setting.

It seems not user-friendly but be acceptable on the first implementation of this Client Conformance Profile.

Of course, it might be better to do auto-updating each client's setting items in comply with FAPI Security Profile setting to be enabled. How about realizing it as follow-up task later?

As for 3, Yes, I agree with you.

BTW, considering other FAPI based security profile support in the future (e.g. UK's OpenBanking Security Profile, Australia's Consumer Data Rights(CDR)), it might be better that we add Security Profile pull-down menu (FAPI Read Only, FAPI Read and Write, ...) instead of FAPI switch.

Summarizing up, my proposal is as follows :

* [PKCE Enforce Setting] Keep "PKCE enforce" switch of Client Settings on Admin Console.

* [Security Profile Change] When the admin changes the client's Security Profile setting (e.g. none -> FAPI RW), keycloak checks whether the current client's settings satisfy the Security Profile to be enabled (e.g. FAPI RW). If not, keycloak shows an appropriate message in order to tell the admin what he should do.

* [Client Setting Change under Security Profile Enabled] When the admin changes some client's settings that does not satisfy the current enabled Security Profile setting, keycloak shows an appropriate message in order to tell the admin what's wrong.

WDYT?

Reply:
I'm not sure if automatic updating settings is a good idea:

* For some settings we would need to guess the option we find best so that we are assuming risks by choosing one.
* Like stated before, some settings have a huge impact o how the client interacts with the server as well as with clients
* FAPI is basically a set of recommendations and obligations in order to provide a "bullet-proof" OpenID deployment. As such, I think we should recommend settings and delegate the actual work to admins (related to first bullet above)

In regards to recommendations, in addition to the policy enforcement shall we also provide a good UI so that we list what a client is missing in order to comply with FAPI?

Reply:
I see, I agree with you. So what about the following?

* when the admin changes the client conformance profile but some settings violates against the new profile, show the separate UI to show the recommended settings per profile.